omod/src/main/java/org/openmrs/module/adminui/page/controller/myaccount/ChangeSecretQuestionPageController.java

```diff
@@ -9,98 +9,72 @@
-                       BindingResult errors,
-                       @SpringBean("userService") UserService userService,
-                       HttpServletRequest request) {
-
-        if (!secretQuestion.getAnswer().equals(secretQuestion.getConfirmAnswer())) {
-            model.put("secretQuestion", secretQuestion);
-            errors.rejectValue("confirmAnswer", "adminui.account.answerAndConfirmAnswer.doesNotMatch");
-            return "myaccount/changeSecretQuestion";
-        }
-
-        return changeSecretQuestion(secretQuestion, userService, request);
-    }
-
-    private String changeSecretQuestion(SecretQuestion secretQuestion, UserService userService, HttpServletRequest request) {
-        try {
-            userService.changeQuestionAnswer(secretQuestion.getPassword(), secretQuestion.getQuestion(), secretQuestion.getAnswer());
-            InfoErrorMessageUtil.flashInfoMessage(request.getSession(), "adminui.account.secretQuestion.success");
-        } catch (Exception ex) {
-            request.getSession().setAttribute(UiCommonsConstants.SESSION_ATTRIBUTE_ERROR_MESSAGE, "registrationapp.save.fail");
-            return "myaccount/changeSecretQuestion";
-        }
-        return "myaccount/myAccount";
-    }
-
-    public class SecretQuestion {
-
-        private String password;
-        private String question;
-        private String answer;
-        private String confirmAnswer;
-
-        public SecretQuestion(String password, String question, String answer, String confirmAnswer) {
-            this.password = password;
-            this.question = question;
-            this.answer = answer;
-            this.confirmAnswer = confirmAnswer;
-        }
-
-        public String getPassword() {
-            return password;
-        }
-
-        public void setPassword(String password) {
-            this.password = password;
-        }
-
-        public String getQuestion() {
-            return question;
-        }
-
-        public void setQuestion(String question) {
-            this.question = question;
-        }
-
-        public String getAnswer() {
-            return answer;
-        }
-
-        public void setAnswer(String answer) {
-            this.answer = answer;
-        }
-
-        public String getConfirmAnswer() {
-            return confirmAnswer;
-        }
-
-        public void setConfirmAnswer(String confirmAnswer) {
-            this.confirmAnswer = confirmAnswer;
-        }
-
-        @Override
-        public String toString() {
-            return "Question=" + question + ", Answer=" + answer;
-        }
-    }
+	
+	public void get(PageModel model) {
+		String secretQuestion = fetchExistingSecretQuestionOrEmpty();
+		model.put("secretQuestion", secretQuestion);
+	}
+	
+	public String post(PageModel model, @RequestParam("password") String password,
+	                   @RequestParam("question") String question, @RequestParam("answer") String answer,
```

Comment:
Setting the secret question and answer should be optional. Also if a user submits the form with some missing fields this will fail